### PR TITLE
[Enhancement] improve join reorder greedy algorithm

### PR DESCRIPTION
**Pull request by Youngwb**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

   StarRocks join reorder algorithm can only generate single sequence left deep tree. such like ABCD four table. A is the table which has the smallest rowcount,  StarRocks will generate only left deep tree start from A, like [ABCD], [ACBD],[ADCB]...

In this PR, we will improve join reorder greedy alogrithm
1. To avoid only generate single sequence left deep tree which tends to lead to local optimality，we can generate multi sequence left deep tree which could start from any table, will generate tree like [ABCD],[BACD],[CADB]...
2. support generate bushy tree

Performance Test：
1. Test tpcds 1t in 1Fe+4BE(16c + 64G) 
total time costs  ：
mian-f931637a6 ：839804ms
greedy improve： 814438ms

1. Test tpcds 1t in 1Fe+8BE(16c + 64G) 
total time costs  ：
mian-509a3b786：899439ms
greedy improve： 854351ms

### CODE REVIEW

**Comment by Youngwb** [comment]
run starrocks_fe_unittest

**Comment by Youngwb** [comment]
run starrocks_fe_unittest

**Comment by Youngwb** [comment]
run starrocks_fe_unittest

**Comment by Youngwb** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 43 / 43 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinReorderGreedy.java|41|41|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinOrder.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinOrder.java [hunk 1]
```diff
@@ -57,6 +58,11 @@ public double getCost() {
 
         double cost = -1L;
         double rowCount = -1L;
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(expr.getOp().hashCode(), leftChildExpr, rightChildExpr);
+        }
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinOrder.java:65]: Why need modify hashCode?

**Review by kangkaisen — DISMISSED** [review]

**Review by Seaven — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Youngwb** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Youngwb** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 43 / 43 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinReorderGreedy.java|41|41|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinOrder.java|2|2|100.00%|[]|

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**